libdwarf/dwarf_init.3
.\" Copyright (c) 2009 Joseph Koshy.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" This software is provided by Joseph Koshy ``as is'' and
.\" any express or implied warranties, including, but not limited to, the
.\" implied warranties of merchantability and fitness for a particular purpose
.\" are disclaimed.  in no event shall Joseph Koshy be liable
.\" for any direct, indirect, incidental, special, exemplary, or consequential
.\" damages (including, but not limited to, procurement of substitute goods
.\" or services; loss of use, data, or profits; or business interruption)
.\" however caused and on any theory of liability, whether in contract, strict
.\" liability, or tort (including negligence or otherwise) arising in any way
.\" out of the use of this software, even if advised of the possibility of
.\" such damage.
.\"
.\" $Id$
.\"
.Dd November 9, 2011
.Dt DWARF_INIT 3
.Os
.Sh NAME
.Nm dwarf_init ,
.Nm dwarf_elf_init
.Nd allocate a DWARF debug descriptor
.Sh LIBRARY
.Lb libdwarf
.Sh SYNOPSIS
.In libdwarf.h
.Ft int
.Fo dwarf_init
.Fa "int fd"
.Fa "int mode"
.Fa "Dwarf_Handler errhand"
.Fa "Dwarf_Ptr errarg"
.Fa "Dwarf_Debug *ret"
.Fa "Dwarf_Error *err"
.Fc
.Ft in
.Fo dwarf_elf_init
.Fa "Elf *elf"
.Fa "int mode"
.Fa "Dwarf_Handler errhand"
.Fa "Dwarf_Ptr errarg"
.Fa "Dwarf_Debug *ret"
.Fa "Dwarf_Error *err"
.Fc
.Sh DESCRIPTION
These functions allocate and return a
.Vt Dwarf_Debug
instance for the object denoted by argument
.Ar fd
or
.Ar elf .
This instance would be used for subsequent access to debugging information in the object by other functions in the DWARF(3) library.
.Pp
For function
.Fn dwarf_init ,
argument
.Ar fd
denotes an open file descriptor referencing a compilation object.
Function
.Fn dwarf_init
implicitly allocates an
.Vt Elf
descriptor for argument
.Ar fd .
.Pp
For function
.Fn dwarf_elf_init ,
argument
.Ar elf
denotes a descriptor returned by
.Xr elf_begin 3
or
.Xr elf_memory 3 .
.Pp
Argument
.Ar mode
specifies the access mode desired.
It should be at least as permissive as the mode with which
the file descriptor
.Ar fd
or the ELF descriptor
.Ar elf
was created with.
Legal values for argument
.Ar mode
are:
.Pp
.Bl -tag -width "DW_DLC_WRITE" -compact
.It DW_DLC_RDWR
Permit reading and writing of DWARF information.
.It DW_DLC_READ
Operate in read-only mode.
.It DW_DLC_WRITE
Permit writing of DWARF information.
.El
.Pp
Argument
.Ar errhand
denotes a function to be called in case of an error.
If this argument is
.Dv NULL
then a default error handling scheme is used.
See
.Xr dwarf 3
for a description of the error handling scheme used by the
DWARF(3) library.
.Pp
Argument
.Ar errarg
is passed to the error handler function denoted by argument
.Ar errhand
when it is invoked.
.Pp
Argument
.Ar ret
points to the memory location that will hold a
.Vt Dwarf_Debug
reference on a successful call these functions.
.Pp
Argument
.Ar err
references a memory location that would hold a
.Vt Dwarf_Error
descriptor in case of an error.
.Ss Memory Management
The
.Vt Dwarf_Debug
instance returned by these functions should be freed using
.Fn dwarf_finish .
.Sh RETURN VALUES
These functions return the following values:
.Bl -tag -width ".Bq Er DW_DLV_NO_ENTRY"
.It Bq Er DW_DLV_OK
This return value indicates a successful return.
.It Bq Er DW_DLV_ERROR
The operation failed.
.It Bq Er DW_DLV_NO_ENTRY
The object specified by arguments
.Ar "fd"
or
.Ar "elf"
did not contain debug information.
.El
.Sh IMPLEMENTATION NOTES
The current implementation does not support access modes
.Dv DW_DLC_RDWR
and
.Dv DW_DLC_WRITE .
.Sh EXAMPLES
To initialize a
.Vt Dwarf_Debug
instance from a open file descriptor referencing an ELF object, and
with the default error handler, use:
.Bd -literal -offset indent
Dwarf_Error err;
Dwarf_Debug dbg;

if (dwarf_init(fd, DW_DLC_READ, NULL, NULL, &dbg, &err) !=
    DW_DLV_OK)
	errx(EXIT_FAILURE, "dwarf_init: %s", dwarf_errmsg(err));
.Ed
.Sh SEE ALSO
.Xr dwarf 3 ,
.Xr dwarf_errmsg 3 ,
.Xr dwarf_finish 3 ,
.Xr dwarf_get_elf 3 ,
.Xr elf_begin 3 ,
.Xr elf_memory 3


RELEASE-NOTES
.. $Id$

.. This file contains a template for use when writing release notes.
.. It needs to be updated with release-specific content prior to
.. cutting a release.  RST comments (such as this one) also need to be
.. removed prior to the release.

.. The tokens '%.*%' need be replaced with actual content.

Release Notes for Elftoolchain Software Version %.%.%
=====================================================

About The Project
-----------------

The `Elftoolchain project`_ develops BSD-licensed implementations of
essential compilation tools and libraries for handling ELF based program
objects.

About The Release
-----------------

.. Describe the rationale for the release (e.g. new features,
.. significant bug fixes, etc.).

Libraries and Utilities
~~~~~~~~~~~~~~~~~~~~~~~

This release comprises the following libraries and utilities.

  =================       ==========================================
  **Name**                **Description**
  =================       ==========================================
  **ar**                  Archive manager.
  **addr2line**           Debug tool.
  **brandelf**            Manage the ELF brand on executables.
  **c++filt**             Translate encoded symbols.
  **elfcopy**             Copy and translate between object formats.
  **elfdump**             Diagnostic tool.
  **findtextrel**         Find undesired text relocations.
  **libdwarf**            DWARF access library.
  **libelf**              ELF access library.
  **mcs**                 Manage comment sections.
  **nm**                  List symbols in an ELF object.
  **ranlib**              Add archive symbol tables to an archive.
  **readelf**             Display ELF information.
  **size**                List object sizes.
  **strings**             Extract printable strings.
  **strip**               Discard information from ELF objects.
  =================       ==========================================

Documentation
~~~~~~~~~~~~~

Each public API and invocable utility has a reference manual entry.
We currently offer %NENTRIES% manual entries, documented in %NFILES%
manual pages.

Additionally, this release contains the following documentation:

  =================       ==========================================
  **Name**                **Description**
  =================       ==========================================
  libelf-by-example       A tutorial introduction to **libelf**.
  =================       ==========================================

Test Suites
~~~~~~~~~~~

The release contains the following test suites:

  =================       ==========================================
  **Name**                **Description**
  =================       ==========================================
  ar                      Test the **ar** utility.
  elfcopy                 Test the **elfcopy** utility.
  elfdump                 Test the **elfdump** utility.
  libdwarf                Test the **libdwarf** library.
  libelf                  Test the **libelf** library.
  nm                      Test the **nm** utility.
  =================       ==========================================


System Requirements
-------------------

.. Hardware and software requirements for using this software.

This software is designed to run on Unix(TM)-like operating systems
such as the BSD-family of operating systems and GNU/Linux.

This release has been built and tested on the following operating
systems:

  ====================  ===========  ===========================
  **Operating System**  **Version**  **Supported Architectures**
  --------------------  -----------  ---------------------------
  `DragonFly BSD`_      2.10.1       i386
  FreeBSD_              10.2         amd64 & i386
  Minix_                3.2.0        i386
  NetBSD_               v5.0.2       i386
  OpenBSD_              v5.0         i386
  Ubuntu_ GNU/Linux     14.04LTS     x86_64
  ====================  ===========  ===========================


Installation and Upgrades
=========================

Installation
------------

Instructions for building and installing this software from source are
described in the file "INSTALL".

Upgrading
---------

.. Special notes about upgrading this software from a prior release.
.. For example, if we introduce any backwards-incompatible behaviour,
.. or if we deprecate existing behaviour.


Release Information
===================

Changes in this release
-----------------------

.. A list of significant changes in the release.

Outstanding Issues
------------------

.. Problems discovered when testing the release.

Known Limitations
-----------------

.. Known limitations.

Test Statistics
---------------

The test summary for this release is presented below:

 =========  =========  ================  ===============  ================  ===================
 **Suite**  **Tests**  **DragonFlyBSD**  **FreeBSD**      **NetBSD**        **Ubuntu 10.04LTS**
 =========  =========  ================  ===============  ================  ===================
 ar
 elfcopy
 elfdump
 libdwarf
 libelf
 nm
 =========  =========  ================  ===============  ================  ===================

Key:

  :P:
      Test successes.
  :F:
      Test failures.
  :U:
      Unresolved tests.

Notes
~~~~~

More Information
================

The project's website is at http://elftoolchain.sourceforge.net/.

Developer Community
-------------------

The project's developers may be contacted using the mailing list:
``<elftoolchain-developers@lists.sourceforge.net>``.

Reporting Bugs
--------------

Please use our `Trac instance`_ for viewing existing bug reports and
for submitting new bug reports.


Copyright and License
=====================

This software is copyright its authors, and is distributed under the
`BSD License`_.

.. _BSD License: http://www.opensource.org/licenses/bsd-license.php
.. _DragonFly BSD: http://www.dragonflybsd.org/

libdwarf-test-suite/ts/dwarf_init/dwarf_init.c
/*-
 * Copyright (c) 2010 Kai Wang
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <assert.h>
#include <dwarf.h>
#include <errno.h>
#include <fcntl.h>
#include <libdwarf.h>
#include <string.h>

#include "driver.h"
#include "tet_api.h"

static void tp_dwarf_init(void);
static struct dwarf_tp dwarf_tp_array[] = {
	{"tp_dwarf_init", tp_dwarf_init},
	{NULL, NULL},
};
#include "driver.c"

static void
tp_dwarf_init(void)
{
	Dwarf_Debug dbg;
	Dwarf_Error de;
	int fd, result;

	result = TET_UNRESOLVED;

	assert(_cur_file != NULL);
	dbg = NULL;
	if ((fd = open(_cur_file, O_RDONLY)) < 0) {
		tet_printf("open %s failed: %s", _cur_file, strerror(errno));
		result = TET_FAIL;
		goto done;
	}

	if (dwarf_init(fd, DW_DLC_READ, NULL, NULL, &dbg, &de) != DW_DLV_OK) {
		tet_printf("dwarf_init failed: %s", dwarf_errmsg(de));
		result = TET_FAIL;
		goto done;
	}

	if (dwarf_init(-1, DW_DLC_READ, NULL, NULL, &dbg, &de) != DW_DLV_ERROR) {
		tet_infoline("dwarf_init didn't return DW_DLV_ERROR when"
		    " called with fd(-1)");
		result = TET_FAIL;
		goto done;
	}

	if (result == TET_UNRESOLVED)
		result = TET_PASS;

done:
	TS_DWARF_FINISH(dbg, de);
	TS_RESULT(result);
}
